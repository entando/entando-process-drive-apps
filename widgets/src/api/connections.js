import { METHODS, DOMAINS } from 'api/constants';

import { CONNECTIONS } from 'mocks/taskList/configs';
import makeRequest from 'api/makeRequest';

// eslint-disable-next-line import/prefer-default-export
export const getConnections = async () =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: CONNECTIONS,
    useAuthentication: false,
  });
