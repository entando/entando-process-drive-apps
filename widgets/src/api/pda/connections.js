import { METHODS, DOMAINS } from '../constants';

import MOCK_CONNECTIONS, {
  TEST_CONNECTION,
  SAVE_CONNECTION,
  DELETE_CONNECTION,
} from '../../mocks/pda/connections';
import makeRequest from '../makeRequest';

export const getConnections = async () =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: MOCK_CONNECTIONS,
    useAuthentication: true,
  });

export const deleteConnection = async name =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: `/connections/${name}`,
    method: METHODS.DELETE,
    mockResponse: DELETE_CONNECTION,
    useAuthentication: true,
  });

export const saveConnection = async connection =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: `/connections`,
    method: METHODS.PUT,
    mockResponse: SAVE_CONNECTION,
    useAuthentication: true,
    body: JSON.stringify(connection),
  });

export const createConnection = async connection =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: `/connections`,
    method: METHODS.POST,
    mockResponse: SAVE_CONNECTION,
    useAuthentication: true,
    body: JSON.stringify(connection),
  });

export const testConnection = async name =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: `/connections/${name}/test`,
    method: METHODS.GET,
    mockResponse: TEST_CONNECTION,
    useAuthentication: true,
  });
