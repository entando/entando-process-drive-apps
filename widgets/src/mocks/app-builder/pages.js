const taskListConfigs = {
  payload: {
    code: 'pda_task_list',
    config: {
      groups:
        '[{"label":"Administrators","key":"Administrators","checked":true},{"label":"broker","key":"broker","checked":true},{"label":"admin","key":"admin","checked":false},{"label":"rest-all","key":"rest-all","checked":true},{"label":"kiemgmt","key":"kiemgmt","checked":false},{"label":"kie-server","key":"kie-server","checked":true},{"label":"appraiser","key":"appraiser","checked":false},{"label":"manager","key":"manager","checked":true},{"label":"supplier","key":"supplier","checked":true},{"label":"IT","key":"IT","checked":true},{"label":"PM","key":"PM","checked":true},{"label":"approver","key":"approver","checked":true}]',
      options:
        '[{"key":"newPageOnClick","label":"Open new page on table row click","checked":false},{"key":"showClaim","label":"Show Claim Button","checked":true},{"key":"showComplete","label":"Show Complete Button","checked":true}]',
      columns:
        '[{"name":"id","position":0,"isVisible":true},{"name":"name","position":1,"isVisible":true},{"name":"createdBy","position":2,"isVisible":true},{"name":"createdAt","position":3,"isVisible":true},{"name":"dueTo","position":4,"isVisible":true},{"name":"status","position":5,"isVisible":true},{"name":"description","position":6,"isVisible":true},{"name":"owner","position":7,"isVisible":true},{"name":"inputData","position":8,"isVisible":true},{"name":"outputData","position":9,"isVisible":true},{"name":"priority","position":10,"isVisible":true},{"name":"subject","position":11,"isVisible":true},{"name":"type","position":12,"isVisible":true},{"name":"form","position":13,"isVisible":true},{"name":"activatedAt","position":14,"isVisible":true},{"name":"skipable","position":15,"isVisible":true},{"name":"workItemId","position":16,"isVisible":true},{"name":"processId","position":17,"isVisible":true},{"name":"processDefinitionId","position":18,"isVisible":true},{"name":"parentId","position":19,"isVisible":true},{"name":"slaCompliance","position":20,"isVisible":true},{"name":"slaDueTo","position":21,"isVisible":true},{"name":"potentialOwners","position":22,"isVisible":true},{"name":"excludedOwners","position":23,"isVisible":true},{"name":"businessAdmins","position":24,"isVisible":true}]',
      knowledgeSource: 'kieStaging',
      process: 'Mortgage_Process.MortgageApprovalProcess@mortgage-process_1.0.1-SNAPSHOT',
    },
  },
  errors: {},
};

const taskDetailsConfigs = {
  payload: {
    code: 'phase_1_widgets_task_details',
    config: {
      containerId: 'evaluation_1.0.0-SNAPSHOT',
      knowledgeSource: 'kieStaging',
      process: 'evaluation@evaluation_1.0.0-SNAPSHOT',
    },
  },
  metaData: { status: 'draft' },
  errors: [],
};

const completionFormConfigs = {
  payload: {
    code: 'phase_1_widgets_completion_form',
    config: {
      containerId: 'evaluation_1.0.0-SNAPSHOT',
      knowledgeSource: 'kieStaging',
      process: 'evaluation@evaluation_1.0.0-SNAPSHOT',
      settings:
        '{"uiSchema":"{\\n  \\"47078d21-7da5-4d3f-8355-0fcd78b09f39\\": {\\n    \\"reason\\": {\\n      \\"ui:widget\\": \\"textarea\\"\\n    },\\n    \\"performance\\": {\\n      \\"ui:widget\\": \\"updown\\"\\n    }\\n  }\\n}"}',
    },
  },
  metaData: { status: 'draft' },
  errors: [],
};

const SummaryCardConfigs = {
  payload: {
    code: 'summary_card',
    config: {
      containerId: 'evaluation_1.0.0-SNAPSHOT',
      knowledgeSource: 'kieStaging',
      process: 'evaluation@evaluation_1.0.0-SNAPSHOT',
      settings: '{"summaryId": "request"}',
    },
  },
  metaData: { status: 'draft' },
  errors: [],
};

export default {
  TASK_LIST: taskListConfigs,
  TASK_DETAILS: taskDetailsConfigs,
  COMPLETION_FORM: completionFormConfigs,
  SUMMARY_CARD: SummaryCardConfigs,
};
