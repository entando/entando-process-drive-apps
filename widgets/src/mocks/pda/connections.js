export default {
  payload: [
    {
      name: 'kieDev',
      host: 'rhpam7-install-kieserver-rhpam7-install-entando.apps.serv.run',
      port: '80',
      schema: 'http',
      app: '/services/rest/server',
      username: 'pamAdmin',
      connectionTimeout: 5000,
      engine: 'pam',
    },
    {
      name: 'kieStaging',
      host: 'rhpam7-install-kieserver-rhpam7-install-entando.apps.serv.run',
      port: '80',
      schema: 'http',
      app: '/services/rest/',
      username: 'pamAdmin',
      connectionTimeout: 60000,
      engine: 'pam',
    },
    {
      name: 'kie2Server',
      host: 'rhpam7-install-kieserver-rhpam7-install-entando.apps.serv.run',
      port: '80',
      schema: 'http',
      app: '/services/rest/',
      username: 'pamAdmin',
      connectionTimeout: 60000,
      engine: 'pam',
    },
    {
      name: 'test',
      host: 'rhpam7-install-kieserver-rhpam7-install-entando.apps.serv.run',
      port: '80',
      schema: 'http',
      app: '/services/rest/',
      username: 'pamAdmin',
      connectionTimeout: 60000,
      engine: 'pam',
    },
  ],
  errors: {},
};
