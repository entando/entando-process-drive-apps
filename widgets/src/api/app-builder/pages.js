import { METHODS, DOMAINS } from 'api/constants';
import WIDGETS_CONFIG from 'mocks/app-builder/widgets';
import WIDGET_CONFIG_MOCKUPS from 'mocks/app-builder/pages';
import makeRequest from 'api/makeRequest';

// used to fetch mock response using frameId, it depends on WIDGETS_CONFIG setup
const getWidgetConfigMockResponse = frameId => {
  const widgetType = Object.keys(WIDGETS_CONFIG).find(
    iteratedType => WIDGETS_CONFIG[iteratedType].frameId === frameId
  );
  return WIDGET_CONFIG_MOCKUPS[widgetType];
};

export const getPageWidget = (pageCode, frameId) =>
  makeRequest({
    domain: DOMAINS.APP_BUILDER,
    uri: `/api/pages/${pageCode}/widgets/${frameId}`,
    method: METHODS.GET,
    mockResponse: getWidgetConfigMockResponse(frameId) || {},
    useAuthentication: true,
    // forceMock: true,
  });

export const putPageWidget = (pageCode, frameId, configs) =>
  makeRequest({
    domain: DOMAINS.APP_BUILDER,
    uri: `/api/pages/${pageCode}/widgets/${frameId}`,
    method: METHODS.PUT,
    body: configs,
    useAuthentication: true,
  });

// TODO: for ease of development - remove when authetication token is managed by wrapper
export const authenticate = async () => {
  const formData = {
    username: 'admin',
    password: 'adminadmin',
    grant_type: 'password',
  };
  return makeRequest({
    domain: DOMAINS.APP_BUILDER,
    uri: '/api/oauth/token',
    method: METHODS.POST,
    mockResponse: {
      access_token: 'MOCKED_ACCESS_TOKEN',
      token_type: 'bearer',
      refresh_token: 'MOCKED_REFRESH_TOKEN',
      expires_in: 3599,
    },
    body: Object.keys(formData)
      .map(key => `${encodeURIComponent(key)}=${encodeURIComponent(formData[key])}`)
      .join('&'),
    useAuthentication: false,
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      Authorization: `Basic ${btoa('appbuilder:appbuilder_secret')}`,
    },
  });
};
