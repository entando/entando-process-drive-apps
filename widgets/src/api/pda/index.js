import makeRequest from 'api/makeRequest';
import { SERVICE, METHODS } from 'api/constants';

import MOCK_CONNECTIONS from 'mocks/pda/connections';
import MOCK_PROCESSES from 'mocks/pda/processes';

export const getConnections = async () =>
  makeRequest({
    domain: SERVICE.URL,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: MOCK_CONNECTIONS,
    useAuthentication: false,
  });

export const getProcess = async connection =>
  makeRequest({
    domain: SERVICE.URL,
    uri: `/connections/${connection}/processes/definitions`,
    method: METHODS.GET,
    mockResponse: MOCK_PROCESSES,
    useAuthentication: false,
  });
