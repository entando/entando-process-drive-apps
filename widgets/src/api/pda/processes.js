import { METHODS, DOMAINS } from 'api/constants';

import MOCK_PROCESSES from 'mocks/pda/processes';
import makeRequest from 'api/makeRequest';

// eslint-disable-next-line import/prefer-default-export
export const getProcesses = async connection =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: `/connections/${connection}/processes/definitions`,
    method: METHODS.GET,
    mockResponse: MOCK_PROCESSES,
    useAuthentication: true,
  });
