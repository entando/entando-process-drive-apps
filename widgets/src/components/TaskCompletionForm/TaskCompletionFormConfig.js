/* eslint-disable no-console */
import React from 'react';
import PropTypes from 'prop-types';
import {
  FormGroup,
  ControlLabel,
  Button,
  HelpBlock,
  Row,
  Col,
  FormControl,
} from 'patternfly-react';

import withAuth from 'components/common/authentication/withAuth';
import { getConnections } from 'api/pda/connections';
import { getProcesses } from 'api/pda/processes';
import { getPageWidget, putPageWidget } from 'api/app-builder/pages';
import ErrorNotification from 'components/common/ErrorNotification';

import 'patternfly-react/dist/css/patternfly-react.css';
import 'patternfly/dist/css/patternfly.css';
import 'patternfly/dist/css/patternfly-additions.css';

class CompletionFormConfig extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      sourceList: [],
      processList: [],
      settings: {
        uiSchema: '{}',
      },
      knowledgeSource: '',
      selectedProcess: '',
      errorMessage: '',
    };

    this.onChangeKnowledgeSource = this.onChangeKnowledgeSource.bind(this);
    this.onChangeProcess = this.onChangeProcess.bind(this);
    this.onChangeUiSchema = this.onChangeUiSchema.bind(this);
    this.closeNotification = this.closeNotification.bind(this);
    this.handleError = this.handleError.bind(this);
    this.handleSave = this.handleSave.bind(this);
  }

  async componentDidMount() {
    try {
      const { frameId, pageCode } = this.props;

      // getting list of Kie server connections
      const sourceList = await getConnections();
      this.setState({ sourceList: sourceList.payload });

      // getting existing configs
      const pageWidgetsConfigs = await getPageWidget(pageCode, frameId, 'COMPLETION_FORM');

      const configs = pageWidgetsConfigs.payload && pageWidgetsConfigs.payload.config;
      if (configs && configs.knowledgeSource) {
        this.onChangeKnowledgeSource(configs.knowledgeSource, () => {
          if (configs.process) {
            this.onChangeProcess(configs.process, () => {
              if (configs.settings) {
                this.setState({
                  settings: JSON.parse(configs.settings),
                });
              }
            });
          }
        });
      }
    } catch (error) {
      this.handleError(error.message);
    }
  }

  onChangeKnowledgeSource(e, cb = () => {}) {
    const knowledgeSource = e.target ? e.target.value : e;
    this.setState({ knowledgeSource });

    getProcesses(knowledgeSource).then(data => {
      this.setState({ processList: data.payload });

      cb();
    });
  }

  onChangeProcess(e, cb = () => {}) {
    const selectedProcess = e.target ? e.target.value : e;
    this.setState({ selectedProcess });

    cb();
  }

  onChangeUiSchema({ target: { value: uiSchema } }) {
    const { settings } = this.state;
    this.setState({ settings: { ...settings, uiSchema } });
  }

  closeNotification = () => {
    this.setState({ errorMessage: '' });
  };

  handleError(errorMessage) {
    this.setState({ errorMessage });
  }

  async handleSave() {
    const { frameId, pageCode, widgetCode } = this.props;
    const { knowledgeSource, selectedProcess, settings } = this.state;
    const [, containerId] = selectedProcess.split('@');

    const body = JSON.stringify({
      code: widgetCode,
      config: {
        knowledgeSource,
        process: selectedProcess,
        containerId,
        settings: JSON.stringify(settings),
      },
    });

    try {
      const response = await putPageWidget(pageCode, frameId, body);
      console.log('Configs got saved', response);
    } catch (error) {
      this.handleError(error.message);
    }
  }

  render() {
    const {
      knowledgeSource,
      sourceList,
      settings,
      processList = [],
      selectedProcess = '',
      errorMessage,
    } = this.state;

    return (
      <div>
        <ErrorNotification message={errorMessage} onClose={this.closeNotification} />
        <form>
          <Row>
            <Col xs={12}>
              <FormGroup controlId="connection">
                <ControlLabel>Knowledge Source</ControlLabel>
                <select
                  className="form-control"
                  value={knowledgeSource}
                  onChange={this.onChangeKnowledgeSource}
                >
                  <option value="">Select...</option>
                  {sourceList.map(source => (
                    <option key={source.name} value={source.name}>
                      {source.name}
                    </option>
                  ))}
                </select>
                <HelpBlock>Select one of the Kie server connections.</HelpBlock>
              </FormGroup>
              <FormGroup controlId="connection">
                <ControlLabel>Process</ControlLabel>
                <select
                  className="form-control"
                  value={selectedProcess}
                  onChange={this.onChangeProcess}
                >
                  <option value="">Select...</option>
                  {processList.map(process => (
                    <option
                      key={`${process['process-id']}@${process['container-id']}`}
                      value={`${process['process-id']}@${process['container-id']}`}
                    >
                      {`${process['process-name']} @ ${process['container-id']}`}
                    </option>
                  ))}
                </select>
                <HelpBlock>Select one BPM Process.</HelpBlock>
              </FormGroup>
            </Col>
          </Row>
          {selectedProcess && (
            <section>
              <legend>Settings</legend>
              <Row>
                <Col xs={12}>
                  <FormGroup bsClass="form-group" controlId="textarea">
                    <ControlLabel bsClass="control-label">UI Schema</ControlLabel>
                    <FormControl
                      bsClass="form-control"
                      componentClass="textarea"
                      value={settings.uiSchema}
                      onChange={this.onChangeUiSchema}
                    />
                  </FormGroup>
                </Col>
              </Row>
              <Row>
                <Col xs={12} className="text-right">
                  <Button bsClass="btn" bsStyle="primary" onClick={this.handleSave}>
                    Save
                  </Button>
                </Col>
              </Row>
            </section>
          )}
        </form>
      </div>
    );
  }
}

CompletionFormConfig.propTypes = {
  frameId: PropTypes.string.isRequired,
  widgetCode: PropTypes.string.isRequired,
  pageCode: PropTypes.string.isRequired,
};

export default withAuth(CompletionFormConfig, ['connection-list', 'process-definition-list']);
