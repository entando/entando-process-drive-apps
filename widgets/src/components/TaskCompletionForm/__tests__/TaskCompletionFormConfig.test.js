import { render, fireEvent, wait } from '@testing-library/react';
import React from 'react';
import 'mocks/i18nMock';

import mockKeycloak from 'mocks/auth/keycloak';
import WIDGET_CONFIGS from 'mocks/app-builder/widgets';
import CONNECTIONS from 'mocks/pda/connections';
import PROCESSES from 'mocks/pda/processes';
import TaskCompletionFormConfig from 'components/TaskCompletionForm/TaskCompletionFormConfig';

import { getConnections } from 'api/pda/connections';
import { getProcesses } from 'api/pda/processes';
import { getPageWidget, putPageWidget } from 'api/app-builder/pages';

mockKeycloak();

jest.mock('api/app-builder/pages');
jest.mock('api/pda/connections');
jest.mock('api/pda/processes');

beforeEach(() => {
  getConnections.mockClear();
  getProcesses.mockClear();
  getPageWidget.mockClear();
  putPageWidget.mockClear();
});

describe('<TaskCompletionFormConfig />', () => {
  it('renders snapshot correctly', async () => {
    getConnections.mockImplementation(() => Promise.resolve(CONNECTIONS));
    getProcesses.mockImplementation(() => Promise.resolve(PROCESSES));
    getPageWidget.mockImplementation(() => Promise.resolve(WIDGET_CONFIGS.COMPLETION_FORM.configs));

    const { container } = render(
      <TaskCompletionFormConfig
        pageCode={WIDGET_CONFIGS.COMPLETION_FORM.pageCode}
        frameId={WIDGET_CONFIGS.COMPLETION_FORM.frameId}
        widgetCode={WIDGET_CONFIGS.COMPLETION_FORM.widgetCode}
      />
    );

    await wait(() => expect(container).toMatchSnapshot());
  });

  it('calls putPageWidget after Save button is pressed', async () => {
    getConnections.mockImplementation(() => Promise.resolve(CONNECTIONS));
    getProcesses.mockImplementation(() => Promise.resolve(PROCESSES));
    getPageWidget.mockImplementation(() => Promise.resolve(WIDGET_CONFIGS.COMPLETION_FORM.configs));
    putPageWidget.mockImplementation(() => {});

    const { findByText } = render(
      <TaskCompletionFormConfig
        pageCode={WIDGET_CONFIGS.COMPLETION_FORM.pageCode}
        frameId={WIDGET_CONFIGS.COMPLETION_FORM.frameId}
        widgetCode={WIDGET_CONFIGS.COMPLETION_FORM.widgetCode}
      />
    );

    const saveButton = await findByText('Save');
    fireEvent.click(saveButton);

    await wait(() => {
      expect(putPageWidget).toHaveBeenCalledTimes(1);
    });
  });
});
