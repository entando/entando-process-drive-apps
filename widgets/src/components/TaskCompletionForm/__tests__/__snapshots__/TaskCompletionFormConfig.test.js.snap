// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskCompletionFormConfig /> renders snapshot correctly 1`] = `
<div>
  <div>
    
    <form>
      <div
        class="row"
      >
        <div
          class="col-xs-12"
        >
          <div
            class="form-group"
          >
            <label
              class="control-label"
              for="connection"
            >
              Knowledge Source
            </label>
            <select
              class="form-control"
            >
              <option
                value=""
              >
                Select...
              </option>
              <option
                value="kieStaging"
              >
                kieStaging
              </option>
              <option
                value="kie2Server"
              >
                kie2Server
              </option>
              <option
                value="test"
              >
                test
              </option>
            </select>
            <span
              class="help-block"
            >
              Select one of the Kie server connections.
            </span>
          </div>
          <div
            class="form-group"
          >
            <label
              class="control-label"
              for="connection"
            >
              Process
            </label>
            <select
              class="form-control"
            >
              <option
                value=""
              >
                Select...
              </option>
              <option
                value="Mortgage_Process.MortgageApprovalProcess@mortgage-process_1.0.0-SNAPSHOT"
              >
                MortgageApprovalProcess @ mortgage-process_1.0.0-SNAPSHOT
              </option>
              <option
                value="evaluation@evaluation_1.0.0-SNAPSHOT"
              >
                Evaluation @ evaluation_1.0.0-SNAPSHOT
              </option>
              <option
                value="itorders.orderhardware@itorders_1.0.0-SNAPSHOT"
              >
                Order for IT hardware @ itorders_1.0.0-SNAPSHOT
              </option>
              <option
                value="Mortgage_Process.MortgageApprovalProcess@mortgage-process_1.0.1-SNAPSHOT"
              >
                MortgageApprovalProcess @ mortgage-process_1.0.1-SNAPSHOT
              </option>
              <option
                value="com.redhat.bpms.examples.mortgage.MortgageApplication@mortgage_1.0"
              >
                MortgageApplication @ mortgage_1.0
              </option>
              <option
                value="itorders-data.place-order@itorders_1.0.0-SNAPSHOT"
              >
                place-order @ itorders_1.0.0-SNAPSHOT
              </option>
            </select>
            <span
              class="help-block"
            >
              Select one BPM Process.
            </span>
          </div>
        </div>
      </div>
      <section>
        <legend>
          Settings
        </legend>
        <div
          class="row"
        >
          <div
            class="col-xs-12"
          >
            <div
              class="form-group"
            >
              <label
                class="control-label"
                for="textarea"
              >
                UI Schema
              </label>
              <textarea
                class="form-control"
                id="textarea"
              >
                {
  "47078d21-7da5-4d3f-8355-0fcd78b09f39": {
    "reason": {
      "ui:widget": "textarea"
    },
    "performance": {
      "ui:widget": "updown"
    }
  }
}
              </textarea>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="text-right col-xs-12"
          >
            <button
              class="btn btn-primary"
              type="button"
            >
              Save
            </button>
          </div>
        </div>
      </section>
    </form>
  </div>
</div>
`;
